src/dealer/DealerWallet.ts

```diff
@@ -0,0 +1,515 @@
+    }
+
+    const btcAmount = usdAmountDiff / btcPrice
+
+    if (depositOrWithdraw) {
+      assert(btcAmount > 0)
+      // amount more than 3x of collateral should not happen
+      // although it could happen the first time the dealer is launched?
+      assert(btcAmount < 3 * usdLiability * btcPrice)
+    }
+
+    return { btcAmount, depositOrWithdraw }
+  }
+
+  static isOrderNeeded({ usdLiability, usdExposure, btcPrice }) {
+    const { ratio } = this.getExposureRatio({ usdLiability, usdExposure })
+
+    let usdOrderAmount
+    let buyOrSell: IBuyOrSell = null
+
+    yamlConfig.hedging.LOW_SAFEBOUND_RATIO_SHORTING
+
+    try {
+      // long (exposed to change in price in BTC)
+      // will loose money if BTCUSD price drops
+      if (ratio < yamlConfig.hedging.LOW_BOUND_RATIO_SHORTING) {
+        const targetUsd = usdLiability * yamlConfig.hedging.LOW_SAFEBOUND_RATIO_SHORTING
+        usdOrderAmount = targetUsd - usdExposure
+        buyOrSell = "sell"
+      }
+
+      // short (exposed to change in price in BTC)
+      // will loose money if BTCUSD price increase
+      else if (ratio > yamlConfig.hedging.HIGH_BOUND_RATIO_SHORTING) {
+        const targetUsd = usdLiability * yamlConfig.hedging.HIGH_SAFEBOUND_RATIO_SHORTING
+        usdOrderAmount = usdExposure - targetUsd
+        buyOrSell = "buy"
+      }
+
+      // else:
+      // we have no, or next to none, exposure to change in price in BTC
+    } catch (err) {
+      throw Error("can't calculate hedging value")
+    }
+
+    const btcAmount = usdOrderAmount / btcPrice
+
+    if (buyOrSell) {
+      assert(btcAmount >= 0)
+      // assert(usdOrderAmount < usdLiability)
+      // TODO: should be reduce only
+    }
+
+    return { btcAmount, buyOrSell }
+  }
+
+  async rebalance({ btcAmount, depositOrWithdraw, logger }) {
+    const currency = "BTC"
+    const sats = btc2sat(btcAmount)
+
+    const metadata = {
+      type: "exchange_rebalance",
+      currency,
+      ...UserWallet.getCurrencyEquivalent({ sats, fee: 0 }),
+    }
+
+    let subLogger = logger.child({ ...metadata, btcAmount, depositOrWithdraw })
+
+    // deposit and withdraw are from the exchange point of view
+    if (depositOrWithdraw === "withdraw") {
+      const memo = `withdrawal of ${btcAmount} btc from ${this.exchange.name()}`
+
+      // TODO: manage the case of
+      // "Account needs to have positive USD balances"
+
+      // getLastOnChainAddress() could be used with whitelisting for more security
+      const address = await this.getLastOnChainAddress()
+
+      // TODO: need a withdrawal password?
+      // FIXME: No fees? event the on-chain fees?
+
+      let withdrawalResult
+
+      subLogger = subLogger.child({ memo, address })
+
+      try {
+        withdrawalResult = await this.exchange.withdraw(currency, btcAmount, address)
+      } catch (err) {
+        const error = "this.exchange.withdraw() error issue"
+        subLogger.error({ withdrawalResult }, error)
```

Comment:
Shouldn't log error here. If it is an expected error we should catch it specifically and translate to a return value otherwise rethrow and log at a higher level.